src/GafferScene/Instancer.cpp

```diff
@@ -192,9 +191,24 @@ class Instancer::EngineData : public Data
 			return it->second;
 		}
 
+		size_t numPrototypes() const
+		{
+			return m_names->outputChildNames()->readable().size();
+		}
+
 		size_t prototypeIndex( size_t pointIndex ) const
 		{
-			return m_indices ? (*m_indices)[pointIndex] : 0;
+			return ( m_indices ? (*m_indices)[pointIndex] : pointIndex ) % numPrototypes();
```

Comment:
> Did you intend to replace : 0 with : pointIndex here? I'm confused because it looks like a break in behaviour compared to before, but I'm not observing the new behaviour I expect to see (cycling pointwise through prototypes if there are no indices).

That's weird, I do see that behaviour here... maybe your test setup wasn't quite right? I think I'd intended that as a fallback; if there's no index primvar then we'll use point order. I do think that's useful (ie. dropping a Seeds on a Plane gives me a round-robin distribution of prototypes). You're right its a change in behaviour and I should have called it out in the PR and the tooltip. I think I lost track of the fact its a change since I did it a month or so ago... I can revert it if you feel strongly about the behaviour change.

Note its not currently throwing for an invalid (as opposed to empty) index primvar name, though it definitely should be.

> I don't know if it has had an impact or not, but moving `% numPrototypes()` here has replaced what was a straight read from a variable with several function calls, for every iteration of a numPoints loop.

That's a bit dramatic. There is only one such loop that's impacted (compute for `prototypeChildNamesPlug`) and several other numPoints loops are not impacted. Still, you're right I haven't profiled and it might well be a regression.

Also, there was never a straight read from a variable. The original code required doing this, which I found incredibly confusing to reason about:

`size_t prototypeIndex = engine->prototypeIndex( i ) % numPrototypes;`

I'll do some profiling to see if I can measure the impact, but I'd argue the new code is considerably clearer; the engine is in charge. Its not much concern since we only have one call-site that needs to do `% numPrototypes`, but at a point before I used `ChildNamesMap` I needed it in multiple places, so it made sense to move it to the engine. Its a const value for each Engine, so I can definitely bake it during `initPrototypes()` to avoid the function call per-point.

Reply:
> I think I'd intended that as a fallback; if there's no index primvar then we'll use point order. I do think that's useful (ie. dropping a Seeds on a Plane gives me a round-robin distribution of prototypes). You're right its a change in behaviour and I should have called it out in the PR and the tooltip. I think I lost track of the fact its a change since I did it a month or so ago... I can revert it if you feel strongly about the behaviour change.

I don't have a particularly strong feeling either way, other than about needing to advertise breaking changes. On the one hand, some round-robin variation might be handy. But on the other I suspect the repeating pattern might not be truly useable, and the existence of _some_ indexing might make it a bit harder to determine that the _specific_ indexing you're after hasn't taken hold. On balance I'm inclined to keep things as they were. Any other thoughts @murraystevenson?

> That's a bit dramatic.

I was simply pointing out something I'd noticed had a chance of being related to the performance regression (and `% numPrototypes` _was_ a straight read). As I said, I don't know if it has had an impact or not. I agree, the code is clearer now, so if it's unrelated to performance, then great.

Reply:
> On balance I'm inclined to keep things as they were.

From an off-line discussion, Murray wasn't too fussed either way, and agreed the round-robin approach is of limited use in production scenarios. 
Reverted to original behaviour for missing indices in 291503188b573538156f8b476a975ae59a87b9fb.

> I don't know if it has had an impact or not

I haven't had time to profile yet, but its simple enough to remove, so I've done that in 35fb9852b76609d60354af143479b893c630359a